go/bind/extension.go

```diff
@@ -0,0 +1,531 @@
+
+	var conn *rpc.Connection
+	handler := newExtensionGregorHandler(gc, nist)
+	if uri.UseTLS() {
+		rawCA := kbCtx.GetEnv().GetBundledCA(uri.Host)
+		if len(rawCA) == 0 {
+			kbCtx.Log.CDebugf(ctx, "getGregorClient: failed to parse CAs: %s", err)
+			return
+		}
+		conn = rpc.NewTLSConnection(rpc.NewFixedRemote(uri.HostPort),
+			[]byte(rawCA), libkb.NewContextifiedErrorUnwrapper(kbCtx),
+			handler, libkb.NewRPCLogFactory(kbCtx),
+			logger.LogOutputWithDepthAdder{Logger: kbCtx.Log}, rpc.ConnectionOpts{})
+	} else {
+		t := rpc.NewConnectionTransport(uri, nil, libkb.MakeWrapError(kbCtx))
+		conn = rpc.NewConnectionWithTransport(handler, t,
+			libkb.NewContextifiedErrorUnwrapper(kbCtx),
+			logger.LogOutputWithDepthAdder{Logger: kbCtx.Log}, rpc.ConnectionOpts{})
+	}
+	defer conn.Shutdown()
+
+	// Make remote successful call on our ad hoc conn
+	return chat1.RemoteClient{Cli: chat.NewRemoteClient(gc, conn.GetClient())}, nil
+}
+
+func restoreName(gc *globals.Context, name string, membersType chat1.ConversationMembersType) string {
+	switch membersType {
+	case chat1.ConversationMembersType_TEAM:
+		if strings.Contains(name, "#") {
+			return strings.Split(name, "#")[0]
+		}
+		return name
+	default:
+		username := gc.GetEnv().GetUsername().String()
+		return name + "," + username
+	}
+}
+
+func ExtensionPostText(strConvID, name string, public bool, membersType int,
+	body string, pusher PushNotifier) (err error) {
+	defer kbCtx.Trace("ExtensionPostText", func() error { return err })()
+	defer func() { err = flattenError(err) }()
+	defer func() { extensionPushResult(pusher, err, strConvID, "message") }()
+
+	gc := globals.NewContext(kbCtx, kbChatCtx)
+	ctx := chat.Context(context.Background(), gc,
+		keybase1.TLFIdentifyBehavior_CHAT_GUI, nil, chat.NewCachingIdentifyNotifier(gc))
+	defer func() {
+		if err == nil {
+			putSavedConv(ctx, strConvID, name, public, membersType)
+		}
+	}()
+
+	convID, err := chat1.MakeConvID(strConvID)
+	if err != nil {
+		return err
+	}
+	sender := chat.NewBlockingSender(gc, chat.NewBoxer(gc),
+		func() chat1.RemoteInterface { return extensionRi })
+	msg := chat1.MessagePlaintext{
+		ClientHeader: chat1.MessageClientHeader{
+			MessageType: chat1.MessageType_TEXT,
+			TlfName:     restoreName(gc, name, chat1.ConversationMembersType(membersType)),
+			TlfPublic:   public,
+		},
+		MessageBody: chat1.NewMessageBodyWithText(chat1.MessageText{
+			Body: body,
+		}),
+	}
+	if _, _, err = sender.Send(ctx, convID, msg, 0, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func extensionPushResult(pusher PushNotifier, err error, strConvID, typ string) {
+	var msg string
+	if err != nil {
+		msg = fmt.Sprintf("We could not send your %s. Please try from the Keybase app.", typ)
+	} else {
+		msg = fmt.Sprintf("Your %s was shared successfully.", typ)
+	}
+	pusher.LocalNotification("extension", msg, -1, "default", strConvID, "chat.extension")
+}
+
+func ExtensionPostJPEG(strConvID, name string, public bool, membersType int,
```

Comment:
(discussed in chat that it might be a good idea to also have ExtensionPostPNG if the original is a PNG file)

Reply:
Filed as another ticket, right now PNG works, but it will have the wrong MIME type (and a JPEG preview).